### PR TITLE
Kafka Broker Triggers receive events from after they are Ready by default

### PR DESCRIPTION
**Pull request by pierDipi**
So far, Kafka Broker Triggers receive events from the time before
they were ready by default, contrary to the eventing spec intent.

https://github.com/knative/specs/blob/main/specs/eventing/control-plane.md#trigger-lifecycle

> When a Trigger becomes associated with a Broker (either due to creating the Trigger or the Broker), the Trigger MUST only set the Ready condition to true after the Broker has been configured to send all future events matching the spec.filter to the Trigger's spec.subscriber. The Broker MAY send some events to the Trigger's spec.subscriber prior to the Trigger's Readycondition being set to true. When a Trigger is deleted, the Broker MAY send some additional events to the Trigger's spec.subscriber after the deletion.

(notice the terms "all future events" and "MAY send some events
... prior to the Trigger's Ready condition"), the spec clearly mean 
only future events should be sent, (allowing only "some prior events" 
because of the possible race between Readiness and actual 
broker setup... )

This behavior differs from the MTBroker behavior, so now
it's more consistent by default.

Users can still switch back to the previous behavior by changing the 
`auto.offset.reset=latest` to `earliest` in config-kafka-broker-data-plane

## Proposed Changes

- [Support initializing offsets in Trigger reconciler](https://github.com/knative-sandbox/eventing-kafka-broker/commit/4471293345dbf4149b7a9d42d0f3b939f4338b70)
- [Set auto.offset.reset to latest by default](https://github.com/knative-sandbox/eventing-kafka-broker/commit/113a804a87dcab674fc2c993b5e5143231257e51)
- [Add E2E test](https://github.com/knative-sandbox/eventing-kafka-broker/commit/db14c34d682a004f1d9950cf4e9c623bddefeae2)

**Release Note**

```release-note
Previously, triggers were receiving events that were sent to a Kafka Broker before they were created.
Now triggers receive events that are sent to the broker after the trigger is ready by default, skipping events that were sent previously.
To switch back to the previous behavior, you can change the `auto.offset.reset=latest` configuration to `earliest` in the `config-kafka-broker-data-plane` ConfigMap.
```

### CODE REVIEW

File: control-plane/config/eventing-kafka-broker/200-controller/500-controller.yaml [hunk 1]
```diff
@@ -69,15 +69,24 @@ spec:
             - name: SOURCE_DATA_PLANE_CONFIG_MAP_NAMESPACE
               value: knative-eventing
 
-            - name: BROKER_DATA_PLANE_CONFIG_MAP_NAME
+            - name: BROKER_CONTRACT_CONFIG_MAP_NAME
```
Comment (aliok) [inline control-plane/config/eventing-kafka-broker/200-controller/500-controller.yaml:72]: Can we do this change in a separate PR? I don't think it is a good idea have this refactoring/renaming in this PR.

**Comment by aliok** [comment]
Can you improve the release note? 

Something like:
```
Triggers were doing ...bla... which was not compliant with the Knative Eventing specification.
Now triggers will receive events that are sent to the broker after the trigger is ready by default, skipping events that were sent previously.
To switch back to the previous behavior, you can change the `auto.offset.reset=latest` configuration to `earliest` in the `config-kafka-broker-data-plane` ConfigMap.
```

**Review by aliok — CHANGES_REQUESTED** [review]

**Comment by aliok** [comment]
Is this a fair statement for what this PR is doing: When `auto.offset.reset=latest`, we set the offset to latest for the consumer group in **control-plane** before consumption starts. 

I couldn't really grasp how this would fix our issue here. We set the offset in control plane when trigger is still not ready. We might set the offset and then wait 5 mins before the trigger is ready.

Maybe instead of writing me a reply here, can you update the PR description?

/assign

File: control-plane/config/eventing-kafka-broker/200-controller/500-controller.yaml (hunk 1 above)
Comment (pierDipi) [reply control-plane/config/eventing-kafka-broker/200-controller/500-controller.yaml:72]: Done in https://github.com/knative-sandbox/eventing-kafka-broker/pull/2602

File: control-plane/pkg/reconciler/broker/broker.go [hunk 2]
```diff
@@ -271,6 +274,9 @@ func (r *Reconciler) reconcileBrokerTopic(broker *eventing.Broker, securityOptio
 
 	statusConditionManager.TopicReady(topicName)
 	logger.Debug("Topic created", zap.Any("topic", topicName))
+
+	broker.Status.Annotations[kafka.TopicAnnotation] = topicName
```
[the remark this answers is not in the document]
Comment (pierDipi) [reply control-plane/pkg/reconciler/broker/broker.go:278]: to read it in the trigger reconciler instead of assuming or duplicating the logic for internal or external topics (or whatever other logic we come up next).

**Comment by pierDipi** [comment]
Waiting for https://github.com/knative-sandbox/eventing-kafka-broker/pull/2602 to go in.

File: control-plane/pkg/reconciler/broker/broker.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (matzew) [reply control-plane/pkg/reconciler/broker/broker.go:278]: If we stash more content on the `status annotation`, you think this is still "OK" ? 

https://github.com/knative-sandbox/eventing-kafka-broker/blob/main/control-plane/pkg/reconciler/broker/broker.go#L503

as the topic name is in theory not on that `kafka-broker-config` CM

**Comment by pierDipi** [comment]
Maven:
```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-shade-plugin:3.3.0:shade (default) on project receiver: Execution default of goal org.apache.maven.plugins:maven-shade-plugin:3.3.0:shade failed: Plugin org.apache.maven.plugins:maven-shade-plugin:3.3.0 or one of its dependencies could not be resolved: Could not transfer artifact org.apache.maven:maven-artifact:jar:3.0 from/to central (https://repo.maven.apache.org/maven2): transfer failed for https://repo.maven.apache.org/maven2/org/apache/maven/maven-artifact/3.0/maven-artifact-3.0.jar: Connection reset -> [Help 1]
```
/test unit-tests_eventing-kafka-broker_main

File: control-plane/config/eventing-kafka-broker/200-controller/500-controller.yaml [hunk 3]
```diff
@@ -78,6 +78,15 @@ spec:
...
 
+            - name: BROKER_DATA_PLANE_CONFIG_CONFIG_MAP_NAME
+              value: config-kafka-broker-data-plane
+            - name: SOURCE_DATA_PLANE_CONFIG_CONFIG_MAP_NAME
+              value: config-kafka-source-data-plane
+            - name: SINK_DATA_PLANE_CONFIG_CONFIG_MAP_NAME
+              value: config-kafka-source-data-plane
+            - name: CHANNEL_DATA_PLANE_CONFIG_CONFIG_MAP_NAME
+              value: config-kafka-source-data-plane
+
```
Comment (aliok) [inline control-plane/config/eventing-kafka-broker/200-controller/500-controller.yaml:89]: I just realized that we need to make adjustments to this approach when we have dataplane configurations per namespace in the future.

With this PR, we always read the dataplaneConfigmap from the system namespace right?

Comment (pierDipi) [reply control-plane/config/eventing-kafka-broker/200-controller/500-controller.yaml:89]: I'm not sure this is just the name, I think we can still require to have this exact name but in their namespace.

afaik, it's not a requirement that we have so far.

Comment (pierDipi) [reply control-plane/config/eventing-kafka-broker/200-controller/500-controller.yaml:89]: however, "delivery profiles" feature is the way to go IMHO https://github.com/knative-sandbox/eventing-kafka-broker/issues/1412

File: control-plane/pkg/kafka/topic.go [hunk 4]
```diff
@@ -31,6 +31,8 @@ const (
 	DefaultTopicNumPartitionConfigMapKey      = "default.topic.partitions"
 	DefaultTopicReplicationFactorConfigMapKey = "default.topic.replication.factor"
 	BootstrapServersConfigMapKey              = "bootstrap.servers"
+
+	TopicAnnotation = "default.topic"
```
Comment (aliok) [inline control-plane/pkg/kafka/topic.go:35]: Can you explain why we need to store the topic in the annotation?

Comment (pierDipi) [reply control-plane/pkg/kafka/topic.go:35]: https://github.com/knative-sandbox/eventing-kafka-broker/pull/2601#discussion_r963534251

File: control-plane/pkg/reconciler/trigger/controller.go [hunk 5]
```diff
@@ -119,6 +126,9 @@ func NewController(ctx context.Context, watcher configmap.Watcher, configs *conf
 		},
 	})
 
+	reconciler.SecretTracker = impl.Tracker
+	secretinformer.Get(ctx).Informer().AddEventHandler(controller.HandleAll(reconciler.SecretTracker.OnChanged))
```
Comment (aliok) [inline control-plane/pkg/reconciler/trigger/controller.go:130]: Did we simply forget this OR we just need this tracking because of the changes here?

File: control-plane/pkg/reconciler/broker/broker.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (pierDipi) [reply control-plane/pkg/reconciler/broker/broker.go:278]: I think it's ok, it's mostly to decouple concerns, broker owns the topic

 (Btw, I also need this for the scaling feature)

File: control-plane/pkg/reconciler/trigger/controller.go (hunk 5 above)
Comment (pierDipi) [reply control-plane/pkg/reconciler/trigger/controller.go:130]: because of the changes here

**Comment by pierDipi** [comment]
Failed https://github.com/knative-sandbox/eventing-kafka-broker/issues/2612 for unrelated error

/test upgrade-tests_eventing-kafka-broker_main

**Comment by pierDipi** [comment]
I rebased

**Review by aliok — APPROVED** [review]
/lgtm
/approve
